### PR TITLE
Chapter 2

### PR DESCRIPTION
**Pull request by timqsh**
### Solutions for Chapter 2

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -336,7 +348,8 @@ from it!
 ghci> :l src/Chapter2.hs
 -}
 subList :: Int -> Int -> [a] -> [a]
-subList = error "subList: Not implemented!"
+subList a b list = if a<0 || b<0 then [] 
+  else take (b-a+1) (drop a list)
```
Comment (chshersh) [inline src/Chapter2.hs]: Your implementation is correct when the right bound is lower than the left. But I think you can implement it more efficiently by adding a separate case to your check :slightly_smiling_face:

File: src/Chapter2.hs [hunk 2]
```diff
@@ -501,7 +515,9 @@ True
 >>> isThird42 [42, 42, 0, 42]
 False
 -}
-isThird42 = error "isThird42: Not implemented!"
+isThird42 :: [Int] -> Bool
+isThird42 (_:_:t:_) = t==42
```
Comment (chshersh) [inline src/Chapter2.hs]: You don't need to compare `t` with 42 separately, you can pattern match on `42` directly in the pattern.

File: src/Chapter2.hs [hunk 3]
```diff
@@ -728,7 +748,7 @@ value of the element itself
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+smartReplicate l = concat (map (\x->replicate x x) l)
```
Comment (chshersh) [inline src/Chapter2.hs]: Great! 👍🏼 
If you want to write this function even more elegantly, you can use the standard function `concatMap` which is a combination of `concat` and `map` as the name suggests 🙂
After that, you could notice, that you would be able to eta-reduce on the last argument :)

File: src/Chapter2.hs [hunk 4]
```diff
@@ -741,7 +761,8 @@ the list with only those lists that contain a passed element.
 
 🕯 HINT: Use the 'elem' function to check whether an element belongs to a list
 -}
-contains = error "contains: Not implemented!"
+contains :: Int -> [[Int]] -> [[Int]]
+contains num ll = filter (elem num) ll
```
Comment (chshersh) [inline src/Chapter2.hs]: Now, after you've mastered eta-reduction, you can apply such a technique to this function as well  🙂 
But your implementation is already great 💯

File: src/Chapter2.hs [hunk 5]
```diff
@@ -842,7 +864,12 @@ list.
 
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
+rotate :: Int -> [a] -> [a]
+rotate num l = subList start end (cycle l)
+  where
+    start = num
+    end = num + (length l) - 1
```
Comment (chshersh) [inline src/Chapter2.hs]: That's a correct solution! And nice way of reusing `subList` function :+1: 

Though, it does a little bit more than work when the given number `num` is big and the list is small. You will cycle through the list a lot. Maybe you can think of a way to optimize this :slightly_smiling_face:

File: src/Chapter2.hs [hunk 6]
```diff
@@ -858,7 +885,9 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [a] -> [a]
+rewind [] = []
+rewind (x:xs) = (rewind xs) ++ [x]
```
Comment (chshersh) [inline src/Chapter2.hs]: Your solution is correct! However, it is slow. In lists, it is quite slow to add anything at the end of the list. That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

**Review by chshersh — APPROVED** [review]
Congrats on finishing Chapter Two!
And well done on the Advanced Task :+1: 

I left my review with some comments and general feedback :slightly_smiling_face:

File: src/Chapter2.hs (hunk 1 above)
Comment (timqsh) [reply src/Chapter2.hs]: I added check for end>start case. Also refactored function using guards and $ application.

File: src/Chapter2.hs (hunk 2 above)
Comment (timqsh) [reply src/Chapter2.hs]: Nice catch :slightly_smiling_face: Fixed it.

File: src/Chapter2.hs (hunk 3 above)
Comment (timqsh) [reply src/Chapter2.hs]: Done :+1:

File: src/Chapter2.hs (hunk 4 above)
Comment (timqsh) [reply src/Chapter2.hs]: eta-reduce everithing ftw :smile:

File: src/Chapter2.hs (hunk 5 above)
Comment (timqsh) [reply src/Chapter2.hs]: I optimized it using `mod` . That way it would not do excessive cycles right?

File: src/Chapter2.hs (hunk 6 above)
Comment (timqsh) [reply src/Chapter2.hs]: I think I got it :slightly_smiling_face: Can you check my new solution using go pattern?

**Comment by timqsh** [comment]
Thanks @chshersh for thorough review. And for great course for learning Hackell :heart:

File: src/Chapter2.hs (hunk 5 above)
Comment (chshersh) [reply src/Chapter2.hs]: That's right! Nice solution :slightly_smiling_face:

**Comment by chshersh** [comment]
@timqsh Your solutions look great! I think this is ready for merge :checkered_flag:

**Pull request merged**